File: blur_detection/blurDetect.py
#!/usr/bin/env python                                        
 

__author__ = "gary"
__copyright__ = "Copyright 2018"
__license__ = ""
__version__ = "0.1.2"
__maintainer__ = ""
__email__ = ""
__status__ = "Development"

import sys
sys.path.append('../../..')


import numpy as np, scipy as sp
import cv2, argparse
import os

SHOW_LAPLACIAN = False
SHOW_TEXT = False

def blurEvaluate(img):
    """
    Calculate the degree of blur.

    Input:
    img: original image from the memory.
    
    Output:
        degree: value of blur.
    """
    # img may not exist.
    degree = -1
    if img is None:
        return degree

    #image scaler

    shape = img.shape
    rows = shape[0]
    cols = shape[1]
    arr_np = np.array([[rows , cols]])
    len_max = arr_np.max()

    scale_thred = 640.0

    if len_max > scale_thred:
        if rows <= cols:
            ratio = scale_thred/cols
            cols_target = int(scale_thred)
            rows_target = int(ratio*rows)
            #print('rows=',rows,' cols=',cols, ' rows_target=',rows_target, ' cols_target=',cols_target)
        else:
            ratio = scale_thred/rows
            cols_target = int(ratio*cols)
            rows_target = int(scale_thred)
            #print('rows=',rows,' cols=',cols, ' rows_target=',rows_target, ' cols_target=',cols_target)      
    
        image_resize = cv2.resize(img, (cols_target, rows_target), cv2.INTER_LINEAR)

    else:

        image_resize = img


    # img exists.
    img_laplacian = cv2.Laplacian(image_resize, cv2.CV_32F, ksize=3)
    #print('image variance=', image.var())
    degree = cv2.meanStdDev(img_laplacian)[1]
    #print degree
    #print('Degree=', sum(degree**2)/3.0)
    return sum(degree**2)/3.0, image_resize ,img_laplacian


def main(args):
    image = cv2.imread(args.image_path, 0)
    
    #print('image shape=', np.shape(image))
    eval = blurEvaluate(image)
    if eval < 10:
        print('-----------------blurred-----------------[eval]:' , eval)
    else:
        print('-----------------clear-----------------[eval]:',eval)
    
    # print('blurEvaluate=', blurEvaluate(image))
    # print('LocalKurtosis')
    # print('image path =',args.image_path)
    # print('image shape =',image.shape)
    # cv2.imshow('original.jpg', image)
    # cv2.imshow('test.jpg', feature.LocalPowerSpectrumSlope(image, 11))
    # cv2.waitKey(0)
    # cv2.destroyAllWindows()

def parse_arguments(argv):
    """Command line parse."""
    parser = argparse.ArgumentParser()
     
    parser.add_argument('--image_path', type=str,
            help='The path of image to be processed.',
            default='/home/akulaku/Project/BlurDetection/original_material/BlurDetect/DiscriminativeBlur/image/motion0001.jpg')
    return parser.parse_args(argv)

def get_image_intensity_mean(image):
    img_array = np.array(image)
    return np.mean(img_array)


def batch_eval_blur(path, output_dir, blur_thred): 

    blur_class_dir  = output_dir + 'blur_class/'
    clear_class_dir = output_dir + 'clear_class/'
    dark_class_dir  = output_dir + 'dark_class/'

    if not os.path.exists(blur_class_dir):
        os.makedirs(blur_class_dir)
    if not os.path.exists(clear_class_dir):
        os.makedirs(clear_class_dir)
    if not os.path.exists(dark_class_dir):
        os.makedirs(dark_class_dir)

    for file in os.listdir(path): 
        whole_file_name = os.path.join(path, file)
        #output_file = os.path.join(output_dir, file) 
        if True == os.path.isfile(whole_file_name):   
            image = cv2.imread(whole_file_name, cv2.IMREAD_COLOR)
            mean = get_image_intensity_mean(image)
            eval, image_resize, img_laplacian = blurEvaluate(image)
            
            if mean < 35:
                text_blur = '[blur]:'+ str(eval[0]) 
                text_mean = '[mean]:' + str(mean)
                print(text_blur + '  ' + text_mean)

                if True == SHOW_TEXT:
                    cv2.putText(image_resize, text_blur, (30, 30), cv2.FONT_HERSHEY_SIMPLEX, 1.0, (0, 255, 255))
                    cv2.putText(image_resize, text_mean, (30, 80), cv2.FONT_HERSHEY_SIMPLEX, 1.0, (0, 255, 255))

                    cv2.putText(image, text_blur, (30, 30), cv2.FONT_HERSHEY_SIMPLEX, 1.0, (0, 255, 255))
                    cv2.putText(image, text_mean, (30, 80), cv2.FONT_HERSHEY_SIMPLEX, 1.0, (0, 255, 255))

                output_dark_file = os.path.join(dark_class_dir, file)
                #cv2.imwrite(output_blur_file, image)

                if True == SHOW_LAPLACIAN:
                    cv2.imwrite(output_dark_file,np.hstack((image_resize, img_laplacian)))
                else:
                    cv2.imwrite(output_dark_file, image)

            else:
                if eval < blur_thred:
                    #print('-----------------blurred-----------------[eval]:' , eval)
                    text_blur = '[blur]:'+ str(eval[0]) 
                    text_mean = '[mean]:' + str(mean)
                    print(text_blur + '  ' + text_mean)

                    if True == SHOW_TEXT:
                        cv2.putText(image_resize, text_blur, (30, 30), cv2.FONT_HERSHEY_SIMPLEX, 1.0, (0, 0, 255))
                        cv2.putText(image_resize, text_mean, (30, 80), cv2.FONT_HERSHEY_SIMPLEX, 1.0, (0, 0, 255))

                        cv2.putText(image, text_blur, (30, 30), cv2.FONT_HERSHEY_SIMPLEX, 1.0, (0, 0, 255))
                        cv2.putText(image, text_mean, (30, 80), cv2.FONT_HERSHEY_SIMPLEX, 1.0, (0, 0, 255))

                    output_blur_file = os.path.join(blur_class_dir, file)
                    #cv2.imwrite(output_blur_file, image)

                    if True == SHOW_LAPLACIAN:
                        cv2.imwrite(output_blur_file,np.hstack((image_resize, img_laplacian)))
                    else:
                        cv2.imwrite(output_blur_file, image)

                else:
                    #print('-----------------clear-----------------[eval]:',eval)
                    text_clear = '[clear]:'+ str(eval[0])
                    text_mean = '[mean]:' + str(mean)
                    print(text_clear + '  ' + text_mean)
                    
                    if True == SHOW_TEXT:
                        cv2.putText(image_resize, text_clear, (30, 30), cv2.FONT_HERSHEY_SIMPLEX, 1.0, (0, 255, 0))
                        cv2.putText(image_resize, text_mean,  (30, 80), cv2.FONT_HERSHEY_SIMPLEX, 1.0, (0, 255, 0))

                        cv2.putText(image, text_clear, (30, 30), cv2.FONT_HERSHEY_SIMPLEX, 1.0, (0, 255, 0))
                        cv2.putText(image, text_mean,  (30, 80), cv2.FONT_HERSHEY_SIMPLEX, 1.0, (0, 255, 0))

                    output_clear_file = os.path.join(clear_class_dir, file)
                    #cv2.imwrite(output_clear_file, image) 
                    #print('image shape:',image.shape, 'img_laplacian shape:',img_laplacian.shape)
                    if True == SHOW_LAPLACIAN:
                        cv2.imwrite(output_clear_file, np.hstack((image_resize, img_laplacian)))               
                    else:
                        cv2.imwrite(output_clear_file, image)


if __name__ == "__main__":
    main(parse_arguments(sys.argv[1:]))
    blur_thred = 450  #500

    path_input = '/home/Workstation/BlurDetection/original_test_0416/cpp_code/image_list/'  
    path_output = '/home/Workstation/BlurDetection/original_test_0416/cpp_code/pyout/'
    batch_eval_blur(path_input, path_output, blur_thred)

File: blur_detection/cpp_version/blurDetect.cpp
#include <iostream>
#include "opencv2/imgproc/imgproc.hpp"
#include "opencv2/highgui/highgui.hpp"
#include <string>
#include <experimental/filesystem>
#include <math.h>

namespace fs = std::experimental::filesystem;

double contrast_measure( const cv::Mat& src)
{
    cv::Mat Gx, Gy;
    cv::Sobel( src, Gx, CV_32F, 1, 0 );
    cv::Sobel( src, Gy, CV_32F, 0, 1 );
    double normGx = cv::norm( Gx );
    double normGy = cv::norm( Gy );
    double sumSq = normGx * normGx + normGy * normGy;
    return static_cast<float>( 1. / ( sumSq / src.size().area() + 1e-6 ));
}

float blurEvaluate(cv::Mat& img)
{
    //check if image exists
    if(!img.data)
    {
        std::cout << "Image may not exist." << std::endl;
        return -1;
    }
    // use Laplacian calculation
    cv::Mat image;
    cv::Laplacian(img, image, CV_32F, 3);

    // display image
    //cv::imshow( "Display window", image);
    //cv::waitKey(0);
     
    cv::Mat var, tmp;
    cv::meanStdDev(image, tmp, var);

    //std::cout << "The rows=" << var.rows << ", the cols=" << var.cols << std::endl;
    double stddev;
    double variance = 0;
    for(int i = 0; i < var.rows; i++)
    {
        stddev = var.at<double>(i, 0);
        variance += stddev * stddev;
        //std::cout << "i=" << i << "=" << stddev << std::endl;
    }

    return variance/3.0;
}

float getImageMeanBrightness(cv::Mat& img)
{
    cv::Mat image_gray, image_float;
    cv::cvtColor(img, image_gray, CV_RGB2GRAY);
    image_gray.convertTo(image_float, CV_32F);
    cv::Scalar scalar = cv::mean(image_float);

    return scalar.val[0];
}

cv::Mat imageResize(cv::Mat& img, float scale_thred)
{
    cv::Mat img_resize;
    float ratio;
    int rows,cols,len_max;
    int rows_target,cols_target;

    rows = img.rows;
    cols = img.cols;
    len_max = (rows >= cols) ? rows : cols;

    //std::cout << "rows:" <<rows << ", " <<" cols :"<< cols << '\n';
    //std::cout << "len_max:" <<len_max << '\n';
    if(len_max > scale_thred)
    {
        if(rows <= cols)
        {
            ratio = scale_thred/cols;
            cols_target = (int)(scale_thred);
            rows_target = (int)(ratio*rows);
        }else
        {
            ratio = scale_thred/rows;
            cols_target = (int)(ratio*cols);
            rows_target = (int)(scale_thred);
        }
        cv::resize(img, img_resize, cv::Size(cols_target, rows_target), CV_INTER_LINEAR);
        //std::cout << "rows:" <<rows << ", " <<" cols :"<< cols << " rows_target :"<< rows_target << " cols_target:" <<cols_target << ", " << '\n';

    }else
    {
        img_resize = img;
    }

    return img_resize;

}

int main(int argc, char** argv)
{
    cv::Mat img, img_resize;

    float mean,blur;
    float mean_thred = 35;
    float blur_thred = 450;
    float scale_thred = 640.0;

    for(auto& p: fs::directory_iterator(argv[1])) {
        
        img = cv::imread(p.path().string());
        img_resize = imageResize(img, scale_thred);
        mean = getImageMeanBrightness(img_resize);
        blur = blurEvaluate(img_resize);

        //std::cout << "rows:" <<img.rows << ", " <<" cols :"<< img.cols << "rows_target:" <<img_resize.rows << ", " <<" cols_target :"<< img_resize.cols << '\n';

        if(mean < mean_thred || blur < blur_thred) {
            std::cout << "[BLUR]:" <<p << ", " <<" blur :"<< blur << "  mean :"<< mean << '\n';

        }else {
            std::cout << "[CLEAR]:" <<p << ", " <<" blur :"<< blur << "  mean :"<< mean << '\n';
        }
    }

    img.release();
}
